kmir/k-src/mir-lexer-syntax.md

```diff
@@ -10,8 +10,12 @@ This module defined the necessary `token` productions.
                       | BBId
                       | DoubleHexDigitNoIntLetter
                       | Whitelisted
+  
+  syntax Whitelisted ::= "main" [token]
 
-  syntax Whitelisted ::= "transmute" | "unwind" | "count"
+  syntax Whitelisted ::= "transmute" [token] | "unwind" [token] | "count" [token]
+
+  syntax Whitelisted ::= "Option" [token] | "None" [token] | "Some" [token] | "unwrap" [token]
```

Comment:
@Baltoli This `Whitelisted` production I think might have been intended for this. I have made the terminals tokens, I am unsure about exactly what happens if I do or don't add the `[token]` attribute. The tutorial seems to indicate it will be beneficial when constructing the AST. Does this result in simpler the KORE in the end?

Reply:
This is a bit subtle and relies on some knowledge of KORE as the backends interpret it. The user manual [documentation](https://github.com/runtimeverification/k/blob/master/docs/user_manual.md#token-attribute) for `[token]` is a bit terse, but it gets to the key point which is that _a token sort is inhabited by domain values and cannot have any constructors declared_. You can think of a "domain value" as the backend's run-time (as opposed to parse-time) analogue of a terminal production in a grammar; they cannot have child nodes, and are more or less represented as a string with some sort information attached.

In KORE, a domain value is written as a special parametric constructor `\dv`, while in K it's `#token`. For example, when you write the boolean literal `true` in K, the backend initially generates the K term `#token("true", "Bool")`, then the KORE term `\dv{SortBool{}}("true")`.

There will be some small differences in the generated KORE depending on whether you make these sorts token sorts or not; I would guess that it's probably marginally better to be pattern-matching on token productions rather than constructors, but I'd need to do some combing through the generated code to confirm that.

In this specific case, I think the most idiomatic way to organise this is as follows:
1. Drop the `Whitelisted` sort entirely unless it's somehow mentioned in the MIR specification; when we see the token `Option` I presume we really do mean that it's an `IdentifierToken` rather than something else.
2. At the point at which you need to mention a specific identifier in a rule or configuration, introduce a token production for that identifier into the relevant semantics module.

For example, [here's](https://github.com/runtimeverification/c-semantics/blob/e3d8a3e812b3344f70719cb954fa4dedb89c4f80/core/semantics/language/exp/primary.k#L23-L43) an example from the C semantics where we write some rules mentioning specific identifiers. Doing it this way keeps the tokens as close to their point of use as possible, and avoids polluting these tokens into the global syntax module (i.e. your grammar as it stands means you can't use `Option` as a variable any more).

Reply:
Thank you, this is a very good explanation.